internal/validator/service.go

```
@@ -13,28 +17,70 @@ type ValidatorService interface {
+	if err != nil {
+		return nil, nil, err
+	}
+
+	justification, err = s.availabilityStore.GetJustification(erasureRoot, shardIndex)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return bundleShard, justification, nil
 }
 
 func (s *validatorService) SegmentShardRequest(ctx context.Context, erasureRoot crypto.Hash, shardIndex uint16, segmentIndexes []uint16) (segmentShards [][]byte, err error) {
-	//TODO implement me
-	panic("implement me")
+	allSegmentsShards, err := s.availabilityStore.GetSegmentsShard(erasureRoot, shardIndex)
+	if err != nil {
+		return nil, err
+	}
+	for _, shardIndex := range segmentIndexes {
+		if len(allSegmentsShards) < int(shardIndex) {
+			return nil, fmt.Errorf("segment shard index %d out of bounds", shardIndex)
+		}
+		segmentShards = append(segmentShards, allSegmentsShards[shardIndex])
+	}
+	return segmentShards, nil
 }
 
 func (s *validatorService) SegmentShardRequestJustification(ctx context.Context, erasureRoot crypto.Hash, shardIndex uint16, segmentIndexes []uint16) (segmentShards [][]byte, justification [][][]byte, err error) {
-	//TODO implement me
-	panic("implement me")
+	segmentShards, err = s.availabilityStore.GetSegmentsShard(erasureRoot, shardIndex)
+	if err != nil {
+		return nil, nil, err
+	}
+	auditShard, err := s.availabilityStore.GetAuditShard(erasureRoot, shardIndex)
+	if err != nil {
+		return nil, nil, err
+	}
+	baseJustification, err := s.availabilityStore.GetJustification(erasureRoot, shardIndex)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for segmentIndex := range segmentShards {
+		auditShardHash := crypto.HashData(auditShard)
+		segmentShardJustification := binary_tree.ComputeTrace(segmentShards, segmentIndex, crypto.HashData)
+
+		// j ⌢ [b] ⌢ T(s, i, H)
+		justification = append(justification, slices.Concat(baseJustification, [][]byte{auditShardHash[:]}, segmentShardJustification))
+	}
+
+	return segmentShards, justification, nil
```

Should we check for empty segmentShards and maybe return an error?

---

idk if returning an error is necessary, but we could optimise a bit so that if we don't get any shards we can return early